### ISSUE TITLE
NCC-2016-017 - Key Loading Script Subject to Man-in-the-Middle Attack

### ISSUE THREAD

**Issue by nathan-at-least** [zcash/zcash#1346]
> Impact: A suitably positioned attacker could tamper with verification and proving keys when
deploying the zcash daemon. At worst, this will cause any transactions created by the zcash
daemon to be rejected by nodes with correct keys.
>
> Description: The fetch-params.sh script retrieves the z9 verification and proving keys from the Zcash
domain. While this connection is made over TLS no certificate validation is performed due
to a known wget discrepancy (see the code comments below). Rather, SHA256 checksums
are computed and checked against the known good values within the script body.
However, keys are copied and linked prior to validation and are not removed if the checksums
does not match. These are loaded upon execution of the Zcash daemon regardless of
prior validation result and could be tampered in transit, effectively bypassing the SHA256
checksums.

We believe this will be mitigated by #1312.

**Comment by daira** [comment]
Was this fixed by #1370?

**Comment by ageis** [comment]
@daira No. I don't understand why we were passing `--no-check-certificate` to wget. Also, what does NCC mean by "a known wget discrepancy"? If we remove [this line](https://github.com/zcash/zcash/blob/master/zcutil/fetch-params.sh#L30), then this will pretty much be fixed.

These were the comments above that choice:
```
        # Note: --no-check-certificate should be ok, since we rely on
        # sha256 for integrity, and there's no confidentiality requirement.
        # Our website uses letsencrypt certificates which are not supported
        # by some wget installations, so we expect some cert failures.
```
I don't think that's reasonable. I will open a PR to enforce certificate verification.

**Comment by daira** [comment]
The problem the report refers to is:  "However, keys are copied and linked prior to validation and are not removed if the checksums does not match." This seems straightforward to fix, by only moving the files from temporary to final locations if the checksums match. Alternatively we can implement #1312, but I don't think we should block on doing that given the very tight timescale.

I am +0 on enabling cert verification, but -1 on relying only on cert verification without fixing the checksum issue.

**Comment by ageis** [comment]
@daira With reference to that description, that was indeed fixed in #1370. However, it was a non-issue even before then. Because we are using `set -e` in the script, it will terminate and refuse to continue when shasum returns a bad exit code (i.e. if the checksums don't match).

**Comment by ageis** [comment]
@daira Let me revise that comment. #1370 fixed the order of operations so that the hashes are checked BEFORE the symlinks are made.

I think we should have cert verification in place anyway.

**Comment by daira** [comment]
So, this issue was already fixed, but ACK e70213103c9d225e714969eb6245d5bb1d9d98fd anyway.
